shogun-lib/src/main/java/de/terrestris/shogun/lib/graphql/controller/BaseGraphQLController.java

```diff
@@ -0,0 +1,185 @@
+import org.springframework.data.history.Revisions;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.lang.reflect.ParameterizedType;
+import java.time.OffsetDateTime;
+import java.util.List;
+import java.util.Optional;
+
+@Log4j2
+public abstract class BaseGraphQLController<E extends BaseEntity, S extends BaseService<? extends BaseCrudRepository<E, Long>, E>> {
+
+    @Autowired
+    protected ObjectMapper objectMapper;
+
+    @Autowired
+    protected S service;
+
+    public List<E> findAll() {
+        return this.service.findAll();
+    }
+
+    public Optional<E> findOne(Long id) {
+        Optional<E> persistedEntity = this.service.findOne(id);
+
+        if (persistedEntity.isEmpty()) {
+            throw new EntityNotAvailableException(String.format("Entity with ID %s is not " +
+                "available", id));
+        }
+
+        return persistedEntity;
+    }
+
+    public Optional<E> findOneForTime(Long id, OffsetDateTime time) {
+        Optional<E> persistedEntity = this.service.findOneByTime(id, time);
+
+        if (persistedEntity.isEmpty()) {
+            throw new EntityNotAvailableException(String.format("Entity with ID %s is not " +
+                "available", id));
+        }
+
+        return persistedEntity;
+    }
+
+    public Revisions<Integer, E> findRevisions(Long id) {
+        Optional<E> persistedEntity = this.service.findOne(id);
+
+        if (persistedEntity.isEmpty()) {
+            throw new EntityNotAvailableException(String.format("Entity with ID %s is not " +
+                "available", id));
+        }
+
+        return this.service.findRevisions(persistedEntity.get());
+    }
+
+    public Optional<Revision<Integer, E>> findRevision(Long id, Integer revId) {
+        Optional<E> persistedEntity = this.service.findOne(id);
+
+        if (persistedEntity.isEmpty()) {
+            throw new EntityNotAvailableException(String.format("Entity with ID %s is not " +
+                "available", id));
+        }
+
+        Optional<Revision<Integer, E>> revision = this.service.findRevision(persistedEntity.get(), revId);
+
+        if (revision.isEmpty()) {
+            throw new EntityNotAvailableException(String.format("Revision %s of entity with ID %s is not " +
+                "available", revId, id));
+        }
+
+        return revision;
+    }
+
+    public List<E> findAllByIds(List<Long> ids) {
+        return this.service.findAllById(ids);
+    }
+
+    public E create(Serializable createEntity) {
+        E entity = this.deserializeInput(createEntity);
+
+        return this.service.create(entity);
+    }
+
+    public E update(Long id, Serializable updateEntity) {
+        Optional<E> persistedEntity = this.service.findOne(id);
+
+        if (persistedEntity.isEmpty()) {
+            throw new EntityNotAvailableException(String.format("Entity with ID %s is not " +
+                "available", id));
+        }
+
+//        updateEntity.put("id", id);
```

Comment:
Line can be removed? Or is it even needed?

Reply:
since this is not updatePartial, id should be already in `updateEntity`